lib/web_ui/lib/src/engine/keyboard_binding.dart

```diff
@@ -0,0 +1,489 @@
+  final Map<int, int> _pressingRecords = <int, int>{};
+
+  // Schedule the dispatching of an event in the future. The `callback` will
+  // invoked before that.
+  //
+  // Returns a callback that cancels the schedule. Disposal of
+  // `KeyBoardConverter` also cancels the shedule automatically.
+  _VoidCallback _scheduleAsyncEvent(Duration duration, ValueGetter<ui.KeyData> getData, _VoidCallback callback) {
+    bool canceled = false;
+    Future<void>.delayed(duration).then<void>((_) {
+      if (!canceled && !_disposed) {
+        callback();
+        dispatchKeyData(getData());
+      }
+    });
+    return () { canceled = true; };
+  }
+
+  // ## About Key guards
+  //
+  // When the user enters a browser/system shortcut (e.g. `cmd+alt+i`) the
+  // browser doesn't send a keyup for it. This puts the framework in a corrupt
+  // state because it thinks the key was never released.
+  //
+  // To avoid this, we rely on the fact that browsers send repeat events
+  // while the key is held down by the user. If we don't receive a repeat
+  // event within a specific duration ([_keydownCancelDuration]) we assume
+  // the user has released the key and we synthesize a keyup event.
+  final Map<int, _VoidCallback> _keyGuards = <int, _VoidCallback>{};
+  // Call this method on the down or repeated event of a non-modifier key.
+  void _startGuardingKey(int physicalKey, int logicalKey, Duration currentTimeStamp) {
+    final _VoidCallback cancelingCallback = _scheduleAsyncEvent(
+      _keydownCancelDuration,
+      () => ui.KeyData(
+        timeStamp: currentTimeStamp + _keydownCancelDuration,
+        change: ui.KeyChange.up,
+        physical: physicalKey,
+        logical: logicalKey,
+        character: null,
+        synthesized: true,
+      ),
+      () {
+        _pressingRecords.remove(physicalKey);
+      }
+    );
+    _keyGuards.remove(physicalKey)?.call();
+    _keyGuards[physicalKey] = cancelingCallback;
+  }
+  // Call this method on an up event event of a non-modifier key.
+  void _stopGuardingKey(int physicalKey) {
+    _keyGuards.remove(physicalKey)?.call();
+  }
+
+  // Parse the HTML event, update states, and dispatch Flutter key data through
+  // [dispatchKeyData].
+  //
+  //  * The method might dispatch some synthesized key data first to update states,
+  //    results discarded.
+  //  * Then it dispatches exactly one non-synthesized key data that corresponds
+  //    to the `event`, i.e. the main key data. The result of this dispatching is
+  //    returned to indicate whether the event is processed by Flutter.
+  //  * Some key data might be synthesized to update states after the main key
+  //    data. They are always scheduled asynchronously with results discarded.
+  bool handleEvent(FlutterHtmlKeyboardEvent event) {
+    final Duration timeStamp = _eventTimeStampToDuration(event.timeStamp!);
+
+    if (_shouldPreventDefault(event)) {
+      event.preventDefault();
+    }
```

Comment:
We can get rid of this if we listen to the response from the framework. The response is a Map that has a `handled` key. If `handled` is `true`, we should call `event.preventDefault()`.

Reply:
Actually, @gspencergoog is doing this in https://github.com/flutter/engine/pull/23655.